msm/src/circuit_design/composition.rs

```diff
@@ -0,0 +1,249 @@
+/// Tools to compose different circit designers.
+use crate::{
+    circuit_design::capabilities::{ColAccessCap, ColWriteCap, LookupCap},
+    columns::ColumnIndexer,
+    logup::LookupTableID,
+};
+use ark_ff::PrimeField;
+
+/// `MPrism` allows one to Something like a Prism, but for Maybe and not just any Applicative.
+///
+/// See
+/// - <https://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Prism.html>
+/// - <https://hackage.haskell.org/package/lens-tutorial-1.0.4/docs/Control-Lens-Tutorial.html>
+pub trait MPrism {
+    /// The lens source type, i.e., the object containing the field.
+    type Source;
+
+    /// The lens target type, i.e., the field to be accessed or modified.
+    type Target;
+
+    fn traverse(&self, source: Self::Source) -> Option<Self::Target>;
+
+    fn re_get(&self, target: Self::Target) -> Self::Source;
+}
+
+pub struct ComposedMPrism<LHS, RHS> {
+    /// The left-hand side of the composition.
+    lhs: LHS,
+
+    /// The right-hand side of the composition.
+    rhs: RHS,
+}
+
+impl<LHS, RHS> ComposedMPrism<LHS, RHS>
+where
+    LHS: MPrism,
+    LHS::Target: 'static,
```

Comment:
Wondering if it might cause an issue. Depending on the data living in Target, we might want to deallocate the value earlier than at the end of the program.

Reply:
Hm. I'll add a comment, but can't comment otherwise now. My hope is that building a circuit will not require more than a (dozen at most!) lenses, so this should be super cheap.